### ISSUE TITLE
false positive with ... and disallow-any-explicit

### ISSUE THREAD

**Issue by KotlinIsland** [python/mypy#9496]
```py
from typing import Callable, TypeVar

Fn = TypeVar("Fn", bound=Callable[..., object])
```
`mypy --disallow-any-explicit .`
```
error: Explicit "Any" is not allowed
    Fn = TypeVar("Fn", bound=Callable[..., object])
    ^
```
But theres no `Any`

I'm guessing the `...` is `Any` but, that's silly.

**Comment by gvanrossum** [comment]
Honestly the disallow-any flag and family were never really battle tested. :-(

Do you want to try and fix it in a PR?

**Comment by KotlinIsland** [comment]
Sure, where do I start?

**Comment by hauntsaninja** [comment]
I touched this code recently! typeanal.py's `analyze_callable_type` is where we turn `Callable[..., object]` into a type :-)

However, I think I'd need more convincing that the current behaviour is silly. `f: Callable[..., X]` is how you'd express `f(*args: Any, **kwargs: Any) -> X`. Eg, `--disallow-any-explicit` catches the following:
```
from typing import Callable

def foo(x: int) -> int: ...

def bar() -> None:
    f: Callable[..., int] = foo
    f("asdf")
```

**Comment by KotlinIsland** [comment]
```py
from typing import Callable, TypeVar

Fn = TypeVar("Fn", bound=Callable[..., object])

def foo(fn: Fn) -> Fn:
    return fn

def bar() -> None: ...

foo(bar)(1)
```
```
error: Too many arguments for "bar"
    foo(bar)(1)
    ^
```
When used as a `TypeVar` it doesn't behave as `Any` and the types are preserved

**Comment by KotlinIsland** [comment]
```py
def foo(fn: Fn) -> None:
    fn(1)
```
Hmm, I guess when used like this it's `Any`.

**Comment by KotlinIsland** [comment]
> ```python
> def foo(fn: Fn) -> None:
>     fn(1)
> ```
> 
> Hmm, I guess when used like this it's `Any`.

That just has the same error about no explicit any `Explicit "Any" is not allowed  [misc]`

**Comment by KotlinIsland** [comment]
@hauntsaninja I agree that it shouldn't be allowed in a type annotation
`f: Callable[..., int] = foo` Umm, can someone say sus much?
But that doesn't look like 'no-explicit-any', it looks more like an 'no-any-generic', or 'incomplete type' or something like that.

But it seems fine (and not actually `Any`) when used as a `bound` in a `TypeVar`. Because the only way it could ever be used would be 

What I want:
```py
from typing import TypeVar, Callable

Fn = TypeVar("Fn", bound=Callable[..., object])

def foo(fn: Fn) -> Fn:  # this is okay, you are only using it in a generic context
    fn(1)  # error, fn's type is generic and unknown
```

**Comment by KotlinIsland** [comment]
This seems very similar to this issue:
```py
from typing import TypeVar, Callable

T = TypeVar("T")

class A:
    def __init__(self, a: int) -> None:
        ...

def foo(t: type[T]) -> T:
    return t()

foo(A)
```

**Comment by KotlinIsland** [comment]
How does `<nothing>` work? would it be applicable here?
When `no-explicit-any` is on `Callable[..., object]` == `Callable[<nothing>, object]`
When `no-explicit-any` is off `Callable[..., object]` == `(*args: Any, **kwargs: Any) -> object`

That seems reasonable to me, wdyt?

**Comment by JanEricNitschke** [comment]
Is there any update on this?
